Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp

```diff
@@ -95,6 +96,139 @@ class NewlineThenWhitespaceStringsTable {
 
 namespace WebCore {
 
+static constexpr unsigned s_maxCachedStringSize = 40000;
+static constexpr unsigned s_minCachedInputSize = 100;
+
+template<typename CharacterType>
+class HTMLFastPathParser;
+
+
+enum class FragmentReuseResult : uint8_t {
+    Success,
+    CannotReuse,
+    ParseFailure
+};
+
+ALWAYS_INLINE static bool isCachedSubtreeValid(Node& cachedContainer)
+{
+    return !cachedContainer.didMutateSubtreeAfterSetInnerHTML();
+}
+
+ALWAYS_INLINE static void invalidateCache(InnerHTMLPrefixCache& cache)
+{
+    cache.cachedContainer.clear();
+    cache.source = String();
+}
+
+template<typename CharacterType>
+static bool isCachedPrefixMatch(const InnerHTMLPrefixCache& cache, std::span<const CharacterType> source, const Element& contextElement)
+{
+    // Source must be at least as long as the cached prefix.
+    if (cache.source.length() > source.size())
+        return false;
+
+    if (cache.contextElementName != contextElement.elementName())
+        return false;
+
+    // Check if character types and prefix content match.
+    StringView cachedStringView(cache.source);
+    if (cachedStringView.is8Bit()) {
+        if constexpr (std::is_same_v<CharacterType, Latin1Character>) {
+            auto cachedSourceSpan = cachedStringView.span8();
+            return std::equal(cachedSourceSpan.begin(), cachedSourceSpan.end(), source.begin());
+        }
+        return false;
+    }
+    if constexpr (std::is_same_v<CharacterType, char16_t>) {
+        auto cachedSourceSpan = cachedStringView.span16();
+        return std::equal(cachedSourceSpan.begin(), cachedSourceSpan.end(), source.begin());
+    }
+    return false;
+}
+
+template<typename CharacterType>
+static FragmentReuseResult tryReuseCachedFragment(std::span<const CharacterType> source, ContainerNode& destinationParent, Element& contextElement)
```

Comment:
This isn't really reusing "cache" since the cache in this case is really a DOM subtree.
I think it's better to call it something like `tryAvoidParsingByCloningExistingSubtree`.